### PR TITLE
Attributes and Errors

### PR DESCRIPTION
**Pull request by awalterschulze**
This pull request contains backwards incompatible changes:

1) Some methods now return errors, instead of panicing.
2) Attributes are checked to be valid attributes.

### CODE REVIEW

File: attr.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+		return Attr(""), fmt.Errorf("%s is not a valid attribute", key)
+	}
+	return a, nil
+}
+
+const (
+	DAMPING            Attr = "Damping"
+	K                  Attr = "K"
+	URL                Attr = "URL"
+	_BACKGROUND        Attr = "_background"
```
Comment (mewmew) [inline attr.go]: _BACKGROUND is not exported, is this intended?

Comment (awalterschulze) [inline attr.go]: Good catch.  What do you think of having these "enums" in all caps?

**Comment by awalterschulze** [comment]
This change also includes changes to the gographviz.Interface to all the user to return errors.

The errInterface is just used to keep track of these errors when a parsed graph is used to populate the gographviz.Interface, `analysed`

**Comment by awalterschulze** [comment]
The errInterface is needed during the analyse step since it is a top down tree transducer and had now way of propagating anything like errors back up.  So to the analyser the interface looks the same, but in the end all the errors are catched and returned to the user by the errInterface.
And since this is never exposed to the user of the API I thought this was an easy solution.

File: attr.go (hunk 1 above)
Comment (awalterschulze) [inline attr.go]: I mean go style is technically to use CamelCase, but I actually prefer the all caps.  What do you think?

**Comment by awalterschulze** [comment]
Maybe I am just misunderstanding you?

**Comment by mewmew** [comment]
Oh, you are right. I must have looked at the previous definition of `Interface`, by clicking on the wrong link in the GitHub UI. After pulling your changes locally, the `Interface` indeed contains errors for the `AddEdge`, `AddNode`, ... methods.

```go
//Interface allows you to parse the graph into your own structure.
type Interface interface {
	SetStrict(strict bool)
	SetDir(directed bool)
	SetName(name string)
	AddPortEdge(src, srcPort, dst, dstPort string, directed bool, attrs map[string]string) error
	AddEdge(src, dst string, directed bool, attrs map[string]string) error
	AddNode(parentGraph string, name string, attrs map[string]string) error
	AddAttr(parentGraph string, field, value string) error
	AddSubGraph(parentGraph string, name string, attrs map[string]string) error
	String() string
}
```
Now I see what made me think it was a bit strange, the errInterface does not have error return values. I would have imagined that you could rid of the errInterface if the Interface contained errors, or that the errInterface could be defined as

```go
type errCatcher struct {
	Interface
	errs []error
}
```

However, a closer look reveals the decision to not have error returns on the errCatcher, simply to not trigger errcheck warnings, since the walker does not check errors during traversal, and you decide to check the error once, after the walk has completed.

This approach definitely works, I just did not see the reason during my first pass of the code.

**Comment by mewmew** [comment]
And indeed, this is precisely what you stated

> The errInterface is needed during the analyse step since it is a top down tree transducer and had now way of propagating anything like errors back up. So to the analyser the interface looks the same, but in the end all the errors are catched and returned to the user by the errInterface.
And since this is never exposed to the user of the API I thought this was an easy solution.

File: attr.go (hunk 1 above)
Comment (mewmew) [inline attr.go]: Had this discussion with my friend the other day, and we were of different opinions. He were accustomed to ALL_CAPS from C. I had started to prefer CamelCase, as the Go project also seems to have moved in this direction recently.

For instance, from [os](https://godoc.org/os#SEEK_SET)

```go
const (
    SEEK_SET int = 0 // seek relative to the origin of the file
    SEEK_CUR int = 1 // seek relative to the current offset
    SEEK_END int = 2 // seek relative to the end
)
//  Deprecated: Use io.SeekStart, io.SeekCurrent, and io.SeekEnd.
```

The ALL_CAPS enums were deprecated (for dependency reasons), and the new enums defined in `io` were CamelCase.

**Comment by mewmew** [comment]
I should try to get better at reading your comments in advance. Downloaded the code first to get my own understanding of how it works, and then read your second reply only after I wrote my reply.

**Comment by awalterschulze** [comment]
Ok cool I have changed all the ALLCAPS Constants to CamelCase.
I think this will prevent a future issue where golint fails on an API level.

**Comment by awalterschulze** [comment]
Now the root folder totally passed golint and I have added error returns on the whole interface.
I hope this will make it easier for users to implement custom graphs.

**Comment by mewmew** [comment]
Looks good. You have a :+1: from me to merge, when you feel satisfied.

**Comment by awalterschulze** [comment]
Thank you for the review.  I really appreciate it.
Here goes a backwards incompatible change for good.

**Pull request merged**